File: system-base/src/main/java/com/levnovikov/system_base/BackStateInteractor.java
package com.levnovikov.system_base;

/**
 * Author: lev.novikov
 * Date: 6/1/18.
 */

import com.levnovikov.system_base.back_handling.BackHandler;
import com.levnovikov.system_base.node_state.ActivityState;

public abstract class BackStateInteractor<R extends Router> extends StateInteractor<R>
        implements BackHandler {

    public BackStateInteractor(R router, ActivityState activityState) {
        super(router, activityState);
        /*
         * If interactor need to handle onBackPressed
         */
        router.setBackHandler(this);
        activityState.addToBackStack(router.getClass());
    }

    @Override
    public boolean isLastInStack(Class<? extends Router> _class) {
        return activityState.isLastInBackStack(_class);
    }

    @Override
    public void popLastInStack() {
        activityState.popLastInBackStack();
    }

    @Override
    public void removeFromBackStack(Class<? extends Router> _class) {
        activityState.removeFromBackStack(_class);
    }
}

File: system-base/src/main/java/com/levnovikov/system_base/BackStateInteractor.kt
package com.levnovikov.system_base

/**
 * Author: lev.novikov
 * Date: 6/1/18.
 */

import com.levnovikov.system_base.back_handling.BackHandler
import com.levnovikov.system_base.node_state.ActivityState

abstract class BackStateInteractor<R : Router>(
        router: R, activityState: ActivityState
) : StateInteractor<R>(router, activityState), BackHandler {

    init {
        /*
         * If interactor need to handle onBackPressed
         */
        @Suppress("LeakingThis")
        router.setBackHandler(this)
        activityState.addToBackStack(router.javaClass)
    }

    override fun isLastInBackStack(_class: Class<out Router>): Boolean {
        return activityState.isLastInBackStack(_class)
    }

    override fun popLastInBackStack() {
        activityState.popLastInBackStack()
    }

    override fun removeFromBackStack(_class: Class<out Router>) {
        activityState.removeFromBackStack(_class)
    }
}
